src/indieweb_utils/parsing/parse.py

```diff
@@ -0,0 +1,49 @@
+import mf2py
+import requests
+from bs4 import BeautifulSoup
+
+
+class RequestError(Exception):
+    pass
+
+
+def get_parsed_mf2_data(parsed_mf2: mf2py.Parser, html: str = "", url: str = ""):
+    """
+    Return or create an mf2py object from a parsed document, a HTML string, and a URL.
+    """
+    if parsed_mf2:
+        return parsed_mf2
+    elif html:
+        return mf2py.parse(doc=html)
+    elif url:
+        return mf2py.parse(url=url)
+
+    raise RequestError("No soup, url, or HTML document provided.")
+
+
+def _get_soup(html: str = "", soup: BeautifulSoup = BeautifulSoup, url: str = "", headers: dict = {}) -> BeautifulSoup:
+    """
+    Return or create a BeautifulSoup object from a parsed document, a HTML string, and a URL.
+    """
+    if html:
+        return BeautifulSoup(html, "html.parser")
+
+    if soup:
```

Comment:
Are there cases where we already have a soup object but would still call `get_soup`?

Reply:
Yes. We call this at any time when we need to determine whether to use a URL, to parse an existing HTML object, or to use a specified BeautifulSoup object. We could wrap calls to this function in an if statement to check whether soup exists before calling the function?

Reply:
Yeah - I think that's what I would do. Check if we have a soup and if we don't, call this function.